### ISSUE TITLE
Is there a way to export inference result to json file in COCO format?

### ISSUE THREAD

**Issue by rhiver** [open-mmlab/mmdetection#5560]
e.g. 
model = init_detector(config_file, checkpoint_file, device='cuda:0')
img = 'path/img.png'
result = inference_detector(model, img)
# TODO: NEED TO EXPORT RESULT AS JSON IN COCO FORMAT HERE
json file should look like:
{
    "images": [
        {"height": x, "width": y, "id": 1, "file_name": "img.png"}
    ],
    "categories": [
        {"supercategory": "c1", "id": 1, "name": "c1"},
        {"supercategory": "c2", "id": 2, "name": "c2"},
        ...
    ],
    "annotations": [
        {"iscrowd": 0, "image_id":1, "bbox": [a1,b1,c1,d1], "segmentation": [[a1,b1,c1,d1,e1,...]], "category_id": 1, "id": 1, "area": x1},
        {"iscrowd": 0, "image_id":1, "bbox": [a2,b2,c2,d2], "segmentation": [[a2,b2,c2,d2,e2,...]], "category_id": 1, "id": 2, "area": x2}, 
        ...
    ]
}

It seems there is a function results2json() in mmdet.datasets.coco.CocoDataset. 
But I'm not really sure if it works. And there is no tutorial about how to use this function. API doc is not clear enough.

Anyone can help to write a demo code?

**Comment by AronLin** [comment]
You can use the `test.py` or other files for the test in `tools` and set:

    --options jsonfile_prefix=$(the_path_you_want_to_save)

Here is the demo: https://github.com/open-mmlab/mmdetection/blob/master/docs/1_exist_data_model.md#test-without-ground-truth-annotations

**Comment by rhiver** [comment]
> You can use the `test.py` or other files for the test in `tools` and set:
> 
> ```
> --options jsonfile_prefix=$(the_path_you_want_to_save)
> ```
> 
> Here is the demo: https://github.com/open-mmlab/mmdetection/blob/master/docs/1_exist_data_model.md#test-without-ground-truth-annotations

Thx for answering. I tried ur suggestion but got 'Segmentation fault (core dumped)' and no json file was generated.

Here is how I use the test.py.
python tools/test.py '/home/ai/envs/mmdetection/work_dirs/my_config.py' '/home/ai/envs/mmdetection/my_model.pth' --format-only --options jsonfile_prefix='/home/ai/envs/tst'

BTW, my model was not trained with COCO dataset but my custom dataset. I just want to convert the result of inference_detector into coco json format instead of exporting a json file. 
Will ur way solve my problem?

**Comment by AronLin** [comment]
If your dataset is not in COCO format, you can not reach your goal by using this command.

CocoDataset.results2json() can dump the results to a json file in COCO format. But you need to provide an annotation file in the same format as the COCO annotation file. The 'images' and 'categories' are necessary, but the ground truth 'annotations' in the file can be empty and it will not be used in formatting results.

However, it is complex to provide such an annotation file. I guess it will be easier to imitate the logic of CocoDataset.results2json()  to write your results2json() in your dataset.

**Comment by rhiver** [comment]
> If your dataset is not in COCO format, you can not reach your goal by using this command.
> 
> CocoDataset.results2json() can dump the results to a json file in COCO format. But you need to provide an annotation file in the same format as the COCO annotation file. The 'images' and 'categories' are necessary, but the ground truth 'annotations' in the file can be empty and it will not be used in formatting results.
> 
> However, it is complex to provide such an annotation file. I guess it will be easier to imitate the logic of CocoDataset.results2json() to write your results2json() in your dataset.

My dataset was not COCO dataset but my annotation was in COCO json format.

I tried to use CocoDataset.results2json(). But it seems I need to call this function by creating a CocoDataset object first. When I tried to create such an object, it says '__init__() missing 2 required positional arguments: 'ann_file' and 'pipeline''.

**Comment by AronLin** [comment]
The ann_file is the path of your annotation file, the pipeline can be set to `[]`.
And if your annotation format is right, it will work.

**Comment by rhiver** [comment]
> The ann_file is the path of your annotation file, the pipeline can be set to `[]`.
> And if your annotation format is right, it will work.

Thx again. 
I tried to write my own results2json(). And I found the segmentation was in the following format. 
It looks like a matrix which has the same size as the image. I guess a True means this pixel belongs to the instance, right?
But is there a way to get the contour of the instance?

'segmentation' = array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])

**Comment by AronLin** [comment]
You can refer to [mmdet/apis/test.py](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/apis/test.py#L100) to encode the masks.

**Comment by rhiver** [comment]
> You can refer to [mmdet/apis/test.py](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/apis/test.py#L100) to encode the masks.

Thx. I encoded it. And it looks like:

{'bbox': [277.8393249511719, 437.6458740234375, 303.7500305175781, 186.0535888671875], 'score': 0.9999446868896484, 'category_id': 4, 'segmentation': {'size': [674, 1201], 'counts': b'dQj51Pe07I101N2N3M3M2N2O0O2N101O001O001O2N2M2O2N1O001O010O1O1O2N100O001O01O010O01O100O1O0010O01O001O001O1O1O2N100O1O1O10O0101O001O0O010O010O01O01001O1O1O0O100O0010O010000000000O010O00010O01000001O1O0O1O10O0100O10001O000O010O01O01O010001O1O1N100O10O01O01000O20O0O010O10O0001O001O4L4L1O001O0010O00100000O1000O010O0001O001O1O1O1O1O001O000010O010003M2M10O010O01O001O001O2N1O001O001O0000100O002N001O1O000010O010O10O2O000O1O010O0000001O00001O00001O00O2N1O2M2N3M3N101N2O0O10000O1O1O1N2O2N1O1O100O1O1O1N2N2O1O1O10000O2O0O2O1N2N2M2O2N100O100O1O1N2N2M3O100O10000O2O0O100O2M2O1O2N1O2O1N1O1O2M2N3M2O1O2N100O2O1Njkh<'}}

from pycocotools import mask
decoded_seg = mask.decode(encoded_seg)

And I used this code to decode it.
